src/emulator/system.c

```diff
@@ -0,0 +1,1801 @@
+        return false;
+    }
+
+    anMode[6] = nSize;
+
+    SYSTEM_CPU(gpSystem)->nTimeRetrace = OSSecondsToTicks(1.0f / 60.0f);
+
+    return true;
+}
+
+bool systemReset(System* pSystem) {
+    s64 nPC;
+    s32 nOffsetRAM;
+    int eObject; // SystemObjectType
+    CpuBlock block;
+
+    pSystem->nAddressBreak = -1;
+
+    if (romGetImage(SYSTEM_ROM(gpSystem), NULL)) {
+        if (!systemSetupGameRAM(pSystem)) {
+            return false;
+        }
+
+        if (!romGetPC(SYSTEM_ROM(gpSystem), (u64*)&nPC)) {
+            return false;
+        }
+
+        block.nSize = 0x100000;
+        block.pfUnknown = NULL;
+        block.nAddress0 = 0x10001000;
+        block.nAddress1 = nPC & 0x00FFFFFF;
+
+        if (!fn_8000A6A4(pSystem, &block)) {
+            return false;
+        }
+
+        if (!cpuReset(SYSTEM_CPU(gpSystem))) {
+            return false;
+        }
+
+        if (!systemSetRamMode(pSystem)) {
+            return false;
+        }
+
+        cpuSetXPC(SYSTEM_CPU(gpSystem), nPC, 0, 0);
+
+        if (!systemSetupGameALL(pSystem)) {
+            return false;
+        }
+
+        for (eObject = 0; eObject < SOT_COUNT; eObject++) {
+            if (pSystem->apObject[eObject] != NULL) {
+                xlObjectEvent(pSystem->apObject[eObject], 0x1003, NULL);
+            }
+        }
+    }
+
+    return true;
+}
+
+static inline bool systemTestClassObject(System* pSystem) {
+    if (xlObjectTest(pSystem, &gClassSystem)) {
+        pSystem->eMode = SM_STOPPED;
+        pSystem->nAddressBreak = -1;
+
+        return true;
+    }
+
+    return false;
+}
+
+bool systemExecute(System* pSystem, s32 nCount) {
+    if (!cpuExecute(SYSTEM_CPU(gpSystem), pSystem->nAddressBreak)) {
+        if (!systemTestClassObject(pSystem)) {
+            return false;
+        }
+
+        return false;
+    }
+
+    if (pSystem->nAddressBreak == SYSTEM_CPU(gpSystem)->nPC) {
+        if (!systemTestClassObject(pSystem)) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool systemCheckInterrupts(System* pSystem) {
+    s32 iException;
+    s32 nMaskFinal;
+    bool bUsed;
+    bool bDone;
+    SystemException exception;
+    CpuExceptionCode eCodeFinal;
+
+    nMaskFinal = 0;
+    eCodeFinal = CEC_NONE;
+    bDone = false;
+    pSystem->bException = false;
+
+    for (iException = 0; iException < ARRAY_COUNT(pSystem->anException); iException++) {
+        if (pSystem->anException[iException] != 0) {
+            pSystem->bException = true;
+
+            if (!bDone) {
+                if (!systemGetException(pSystem, iException, &exception)) {
+                    return false;
+                }
+
+                bUsed = false;
+
+                if (exception.eCode == CEC_INTERRUPT) {
+                    if (cpuTestInterrupt(SYSTEM_CPU(gpSystem), exception.nMask) &&
+                        ((exception.eTypeMips == MIT_NONE) ||
+                         miSetInterrupt(SYSTEM_MI(gpSystem), exception.eTypeMips))) {
```

Comment:
nit: redundant parens (in this case I think it makes the condition a bit harder to read, the important parens are the ones around the (x || y)